### PR TITLE
add torchcomms backend for xpu support

### PR DESCRIPTION
**Pull request by jemitche1**
This PR adds a torchcomms backed device communicator path for vllm on XPU. The new TorchCommsCommunicator integrates with vllm's existing DeviceCommunicator and routes collectives via torchcomms/XCCL while preserving the existing XPU communicator as the default path.

### CODE REVIEW

**Comment by yma11** [comment]
@jemitche1 can you explain why introduce this communication backend to XPU and any benefit cases?

**Comment by jemitche1** [comment]
@yma11 We are introducing torchcomms support for xpu in vllm primarily to enable the initial integration path as torch's c10d communication layer is expected to be deprecated. Initially, we want to establish functionality, feature and performance parity with c10d.

**Comment by jemitche1** [comment]
Can someone please add the 'ready' or 'verified' label so CI can run?

File: vllm/distributed/parallel_state.py [hunk 1]
```diff
@@ -477,17 +477,29 @@ def __init__(
...
-                current_platform.get_device_communicator_cls()
+            device_comm_qualname = current_platform.get_device_communicator_cls()
+
+            logger.info(
+                "[verify] selected device communicator class: %s",
+                device_comm_qualname,
             )
+
+            device_comm_cls = resolve_obj_by_qualname(device_comm_qualname)
+
```
Comment (frost-intel) [inline vllm/distributed/parallel_state.py:490]: Seems unnecessary.

File: vllm/distributed/parallel_state.py [hunk 2]
```diff
@@ -477,17 +477,29 @@ def __init__(
...
             self.device_communicator = device_comm_cls(
                 cpu_group=self.cpu_group,
                 device=self.device,
                 device_group=self.device_group,
                 unique_name=self.unique_name,
             )
 
+            logger.info(
+                "[verify] initialized device communicator: %s",
```
Comment (frost-intel) [inline vllm/distributed/parallel_state.py:499]:
```suggestion
            logger.debug(
                "Initialized device communicator: %s",
```

File: vllm/platforms/xpu.py [hunk 3]
```diff
@@ -416,13 +416,24 @@ def is_data_center_gpu(cls) -> bool:
...
+            "1",
+            "true",
+            "yes",
+            "on",
+        ):
+            return (
+                "vllm.distributed.device_communicators."
+                "torchcomms_communicator.TorchCommsCommunicator"
+            )
+
```
Comment (frost-intel) [inline vllm/platforms/xpu.py:429]: Use convention in `vllm/vllm/envs.py` instead of `os.environ`.

File: vllm/distributed/device_communicators/torchcomms_communicator.py [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+
+from __future__ import annotations
+
+import os
+from typing import Any
+
+import torch
+import torch.distributed as dist
+import torchcomms
+from torchcomms import ReduceOp
```
Comment (frost-intel) [inline vllm/distributed/device_communicators/torchcomms_communicator.py:12]: You should be using a lazy init, guarding on torchcomms availability and warn then fallback if torchcomms env flag is set but torchcomms isn't available. This will crash for a system without torchcomms installed.

File: vllm/distributed/device_communicators/torchcomms_communicator.py [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+            "Set VLLM_TORCHCOMMS_BACKEND explicitly."
+        )
+
+    if device.type == "xpu":
+        return "xccl"
+
+    raise RuntimeError(
+        f"Could not infer torchcomms backend for device={device}. "
+        "Set VLLM_TORCHCOMMS_BACKEND explicitly."
+    )
```
Comment (frost-intel) [inline vllm/distributed/device_communicators/torchcomms_communicator.py:35]: See `vllm/vllm/platforms/__init__.py`, you already have the current platform programmatically, don't require a check here.

File: vllm/distributed/device_communicators/torchcomms_communicator.py [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+
+    def _log_tensor(self, op: str, tensor: torch.Tensor) -> None:
+        logger.info(
+            "[torchcomms] rank=%s op=%s shape=%s dtype=%s device=%s",
+            self.group_rank,
+            op,
+            tuple(tensor.shape),
+            tensor.dtype,
+            tensor.device,
+        )
```
Comment (frost-intel) [inline vllm/distributed/device_communicators/torchcomms_communicator.py:238]: Logging every collective to info is unusable.

File: vllm/distributed/device_communicators/torchcomms_communicator.py [hunk 7]
```diff
@@ -0,0 +1,409 @@
...
+
+    def all_reduce(self, input_: torch.Tensor) -> torch.Tensor:
+        self._log_tensor("all_reduce:start", input_)
+
+        if self.group_world_size == 1:
+            self._log_tensor("all_reduce:done_world_size_1", input_)
+            return input_
```
Comment (frost-intel) [inline vllm/distributed/device_communicators/torchcomms_communicator.py:245]: You shouldn't even enter this code with a world size == 1, so checks aren't necessary.

**Review by frost-intel — CHANGES_REQUESTED** [review]
I think the path to a mergeable PR is:
0. torchcomms is being integrated into upstream pytorch, if this could be done it would allow your comm to be a ProcessGroup
1. Using ProcessGroup as a base class for should allow you to remove all the env variables which are being set here so you stop mutating os.environ
2. Strip all the debug logging
3. Ensure the base api (all the collectives) are actually working with the torchcomms backend.

**Comment by jemitche1** [comment]
@frost-intel , Thanks for the detailed feedback. I'll address these cleanup items in the next revision. 

I agree that the cleaner long term design is for torchcomms to integrate via pytorch distributed as a process group path. Once that is available, vllm should not mange a separate torchomms communicator directly. For this PR, I am treating the current implementation as an incremental path for torchcomms/xccl in vllm, while keeping the design aligned with eventually moving to process group integration.

File: vllm/distributed/device_communicators/torchcomms_communicator.py [hunk 8]
```diff
@@ -0,0 +1,409 @@
...
+
+
+def _setup_backend_specific_env(
+    *,
+    backend: str,
+    local_rank: int,
+    local_world_size: int,
+) -> None:
+    if backend == "xccl":
+        os.environ.setdefault("CCL_PROCESS_LAUNCHER", "none")
```
Comment (newtdms) [inline vllm/distributed/device_communicators/torchcomms_communicator.py:76]: We should remove any CCL flags from here. It should be part of the launch script as some of it are specific to target platforms/system.

File: vllm/distributed/device_communicators/torchcomms_communicator.py [hunk 9]
```diff
@@ -0,0 +1,409 @@
...
+            group_world_size=self.group_world_size,
+        )
+
+        _setup_backend_specific_env(
+            backend=self.backend,
+            local_rank=local_rank,
+            local_world_size=local_world_size,
+        )
+
+        logger.info(
```
Comment (newtdms) [inline vllm/distributed/device_communicators/torchcomms_communicator.py:127]: May be we can use logger.debug as Frost suggested?

File: vllm/distributed/parallel_state.py (hunk 1 above)
Comment (jemitche1) [reply vllm/distributed/parallel_state.py:490]: Agreed. This is only added for verification. I'll remove the selected communicator logger.info() and restore the simpler  resolve_obj_by_qualname(current_platform.get_device_communicator_cls()) path

File: vllm/distributed/parallel_state.py (hunk 2 above)
Comment (jemitche1) [reply vllm/distributed/parallel_state.py:499]: Agreed. I'll change this to logger.debug() with the suggested wording or remove it if it is not needed after cleanup

File: vllm/platforms/xpu.py (hunk 3 above)
Comment (jemitche1) [reply vllm/platforms/xpu.py:429]: Agreed. I'll remove VLLM_USE_TORCHCOMMS and VLLM_TORCHCOMMS_BACKEND to vllm/envs.py and update xpu.py to use envs.VLLM_USE_TORCHCOMMS instead of parsing via os.env directly

File: vllm/distributed/device_communicators/torchcomms_communicator.py (hunk 5 above)
Comment (jemitche1) [reply vllm/distributed/device_communicators/torchcomms_communicator.py:35]: Agreed. Since this communicator is selected from the xpu platform path, the device-type check here is redundant. I’ll remove _infer_torchcomms_backend() and rely on the platform selection to choose this communicator only for xpu.

For now, the communicator can default to xccl, with VLLM_TORCHCOMMS_BACKEND available as an explicit override, if needed.

File: vllm/distributed/device_communicators/torchcomms_communicator.py (hunk 6 above)
Comment (jemitche1) [reply vllm/distributed/device_communicators/torchcomms_communicator.py:238]: Agreed. I'll remove collective logging info but this was to ensure collectives were being reached (for verification)
